scvi/external/cell2location/_base.py

```diff
@@ -0,0 +1,1060 @@
+                }
+
+            else:
+                args, kwargs = self.module._get_fn_args_from_batch(tensor_dict)
+                args = [a.to(device) for a in args]
+                kwargs = {k: v.to(device) for k, v in kwargs.items()}
+                self.to_device(device)
+
+                means_ = self.module.guide.quantiles([q], *args, **kwargs)
+                means_ = {
+                    k: means_[k].cpu().detach().numpy()
+                    for k in means_.keys()
+                    if k in self.module.model.list_obs_plate_vars()["sites"]
+                }
+                means = {
+                    k: np.concatenate(
+                        [means[k], means_[k]], axis=list(obs_plate.values())[0]
+                    )
+                    for k in means.keys()
+                }
+            i += 1
+
+        # sample global parameters
+        for tensor_dict in train_dl:
+            args, kwargs = self.module._get_fn_args_from_batch(tensor_dict)
+            args = [a.to(device) for a in args]
+            kwargs = {k: v.to(device) for k, v in kwargs.items()}
+            self.to_device(device)
+
+            global_means = self.module.guide.quantiles([q], *args, **kwargs)
+            global_means = {
+                k: global_means[k].cpu().detach().numpy()
+                for k in global_means.keys()
+                if k not in self.module.model.list_obs_plate_vars()["sites"]
+            }
+            break
+
+        for k in global_means.keys():
+            means[k] = global_means[k]
+
+        self.module.to(device)
+
+        return means
+
+    def _posterior_quantile(self, q: float = 0.5, use_gpu: bool = True):
+        """
+        Compute median of the posterior distribution of each parameter pyro models trained without amortised inference.
+
+        Parameters
+        ----------
+        q
+            quantile to compute
+        use_gpu
+            Bool, use gpu?
+
+        Returns
+        -------
+        dictionary {variable_name: posterior median}
+
+        """
+
+        self.module.eval()
+        gpus, device = parse_use_gpu_arg(use_gpu)
+
+        if self.module.model.batch_size is None:
+            args, kwargs = self.module.model._get_fn_args_full_data(self.adata)
+            args = [a.to(device) for a in args]
+            kwargs = {k: v.to(device) for k, v in kwargs.items()}
+            self.to_device(device)
+        else:
+            train_dl = AnnDataLoader(
+                self.adata, shuffle=False, batch_size=self.module.model.batch_size
+            )
+            # sample global parameters
+            for tensor_dict in train_dl:
+                args, kwargs = self.module._get_fn_args_from_batch(tensor_dict)
+                args = [a.to(device) for a in args]
+                kwargs = {k: v.to(device) for k, v in kwargs.items()}
+                self.to_device(device)
+                break
+
+        means = self.module.guide.quantiles([q], *args, **kwargs)
```

Comment:
our `automovedata` decorator would work here

Reply:
though there's a chance it needs to be tweaked.

Overall though, it removes all this device specific code

Reply:
Could I ask you to point me to usage example?